### ISSUE TITLE
fREM not setting random state of cv

### ISSUE THREAD

**Issue by jeromedockes** [nilearn/nilearn#2438]
By default the fREMClassifier and fREMRegressor use a stratified shuffle split
without setting the random seed.

we should either:
   - offer the possibility to set the random state in the fREM objects' interface
   - pass a cross-validator with a set random state rather than an int whenever we use fREM in tests

otherwise it can cause spurious failures like [this one](https://dev.azure.com/Parietal/Nilearn/_build/results?buildId=1265&view=logs&j=c106e126-660f-5fbd-4f14-98be518a1706&t=63088eea-9651-5779-603f-44aaa46b8692)

@thomasbazeille  what do you think is best?

**Comment by thomasbazeille** [comment]
Good catch sry about that.

I guess we should seed the default creation to make results replicable for users and avoid this kind of issue, if someone want to use a different seed (which should not matter too much with high cv number supposedly used) he'll always have the option to pass his cross-validator.

Does it feel right ? If so I can send the PR.

**Comment by jeromedockes** [comment]
>
> I guess we should seed the default creation to make results replicable for users and avoid this kind of issue, if someone want to use a different seed (which should not matter too much with high cv number supposedly used) he'll always have the option to pass his cross-validator.
>
> Does it feel right ? If so I can send the PR.

Another option would be to add a random_state argument, None by default.
Scikit-learn estimators that use randomness such as random forest or
bagging or random projections seem to follow this approach.

**Comment by bthirion** [comment]
> Another option would be to add a random_state argument, None by default. Scikit-learn estimators that use randomness such as random forest or bagging or random projections seem to follow this approach.

I like the approach. Users might wonder what a CV objects should be like, while `random_state` is simpler

**Comment by thomasbazeille** [comment]
I disagree with both arguments. I think fixing a defaut seed is a good idea to make default results reproducible.

I guess the question we should agree on is who is going to want to change this parameters. For me close to noone, when you do 50 ShuffleSplit and bag results you don't care about the seed. So adding a supplementary parameter is not worth it

> Another option would be to add a random_state argument, None by default.
Scikit-learn estimators that use randomness such as random forest or
bagging or random projections seem to follow this approach.

fREM (and Decoder) are no estimator, they are pipelines, and people are not supposed to run several different of them in general since they encapsulate CV, bagging, preprocessing... 

Moreover it will duplicate the parameter of cv Object (override it ?). I think we'd better document in doc that default cv is seeded and how to change it if needed.

**Comment by bthirion** [comment]
So if I understand correctly,  you propose to pass a cross-validator ? This sounds valid to me. 

I was thinking that it was good to replicate the argumennts of sklearn objects, but maybe you're right.

**Comment by jeromedockes** [comment]
> Moreover it will duplicate the parameter of cv Object (override it ?).

that's true. In this case let's do what you said i.e. just set the seed
when cv is an int

> I think we'd better document in doc that default cv is seeded and how to change it if needed.

I don't think you need to add much, maybe just put in the docstring
description of the cv parameter that the stratifiedshufflesplit when cv
is an int uses a fixed seed but even that may not be necessary

**Comment by thomasbazeille** [comment]
> So if I understand correctly, you propose to pass a cross-validator ? This sounds valid to me

Sorry for being unclear. By default, we initialize to fREM(cv=30) parameter which creates :
```
if isinstance(cv, int):
            cv = ShuffleSplit(cv)/ StratifiedShuffleSplit(cv)
```

I'm proposing to seed this line. Users can still override it as before by feeding cv a generator/cv object.


@jeromedockes For the Decoder classes though it's not as easy. We use sklearn [check_cv](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.check_cv.html) helper. That usually creates KFold/StratifiedKFold(cv) when cv is an int but we have to add additional cases if we want to seed the default. 
Do you think it's worth it as well?

**Comment by GaelVaroquaux** [comment]
> I'm proposing to seed this line. Users can still override it as before by feeding cv a generator/cv object.

Sounds right!

This is an issue that does not require expert knowledge of nilearn. I suggest that you explain in details here what needs to be done, but not do it yourself. Rather, we should mark it as something that needs help during the sprint, and you could mentor the person doing it.

**Comment by jeromedockes** [comment]
> @jeromedockes For the Decoder classes though it's not as easy. We use sklearn [check_cv](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.check_cv.html) helper. That usually creates KFold/StratifiedKFold(cv) when cv is an int but we have to add additional cases if we want to seed the default. 
> Do you think it's worth it as well?

Sorry I'm not sure I understand. By default the Decoder will use KFold,
StratifiedKFold or LeaveOneGroupOut and none of these need a random
seed, no?

**Comment by thomasbazeille** [comment]
> This is an issue that does not require expert knowledge of nilearn. I suggest that you explain in details here what needs to be done, but not do it yourself. Rather, we should mark it as something that needs help during the sprint, and you could mentor the person doing it.

I think this was linked to spurious CI failures, so I guess we want to get it out of the way before the sprint ?

> Sorry I'm not sure I understand. By default the Decoder will use KFold,
StratifiedKFold or LeaveOneGroupOut and none of these need a random
seed, no?

Right the default don't shuffle, forget it, I'll send a PR
